File: Ahorn/entities/VirusCloud.jl
module VirusCloud

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/VirusCloud" VirusCloud(x::Integer, y::Integer, damageRate::Number=0.5, health::Integer=3, moveSpeed::Integer=40, spreadRadius::Integer=100)

const placements = Ahorn.PlacementDict(
    "normal" => Ahorn.EntityPlacement(VirusCloud)
)

function Ahorn.selection(entity::VirusCloud)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::VirusCloud, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: framework/extracted/Loenn/entities/virusCloud.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local virusCloud = {}

virusCloud.name = "MaggyHelper/VirusCloud"
virusCloud.depth = -50
virusCloud.fieldInformation = {
    health = {
        minimumValue = 1
    },
    spreadRadius = {
        minimumValue = 0
    },
    moveSpeed = {
        minimumValue = 0
    },
    damageRate = {
        minimumValue = 0
    }
}
virusCloud.placements = {
    {
        name = "normal",
        data = {
            health = 3,
            spreadRadius = 100,
            moveSpeed = 40,
            damageRate = 0.5
        }
    }
}

function virusCloud.sprite(room, entity)
    local sprite = drawableSprite.fromTexture("characters/virus_cloud/dormant", entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 0.5)
    sprite.color = {0.5, 0.0, 0.5, 0.7}
    return sprite
end

function virusCloud.selection(room, entity)
    return utils.rectangle(entity.x - 20, entity.y - 20, 40, 40)
end

return virusCloud
